modules/apps/frontend-js/frontend-js-state-web/src/main/resources/META-INF/resources/deepFreeze.ts

```diff
@@ -0,0 +1,28 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+export default function deepFreeze<T>(value: T): Immutable<T> {
+	if (process.env.NODE_ENV === 'development') {
```

Comment:
Yep. This is a "secret weapon" that we used on Relay. You can program in an _immutable style_ (enforced by the type system) without paying the overhead of using an _actual immutable data-type implementation_. The `'development'` checks are an added layer of protection, so that when people break the rules from a non-TypeScript part of the codebase, there is still a reasonable chance that we'll be able to catch the mistake before it gets too far.